File: app/lib/dominio/actividades.dart
import '../contenido/dictados.dart';
import '../contenido/matematicas.dart';
import '../correccion/dictado.dart';
import '../correccion/tanda.dart';
import '../voz/frases.dart';
import '../voz/locutora.dart';
import 'asignaturas.dart';
import 'guion.dart';

/// Cuántas faltas se repasan de viva voz antes de que el repaso canse más de lo
/// que enseña.
const int maxFaltasARepasar = 5;

// ---------------------------------------------------------------- dictado ---

Guion guionDictado(Dictado dictado) {
  final pasos = <Paso>[
    Habla(Frases.dictadoIntro(dictado.titulo)),
    const Espera(Frases.prepararPapel, [Comando.listo]),
    const Habla(Frases.empezamos),
  ];

  final frases = dictado.frasesDictadas;
  for (var i = 0; i < frases.length; i++) {
    pasos.add(Fragmento(
      indice: i,
      texto: frases[i],
      veces: vecesPorFrase,
    ));
  }

  pasos.add(const Revisar(Frases.dictadoFin));

  return Guion(
    asignatura: Asignatura.dictado,
    titulo: dictado.titulo,
    pasos: pasos,
    // Durante un dictado el niño manda: repite la frase las veces que quiera,
    // ajusta el ritmo y pasa a la siguiente cuando la tiene escrita.
    comandosGlobales: const [
      Comando.repite,
      Comando.masDespacio,
      Comando.masRapido,
      Comando.continua,
    ],
    velocidadInicial: velocidadDeDictado(dictado.nivel),
  );
}

/// Repaso de un dictado que el niño acaba de corregirse él mismo.
///
/// Se le explica la regla de cada palabra que ha tachado: esa es la parte que
/// enseña. Como las ha marcado sobre el texto, no hay nada que adivinar.
Guion guionRepasoDictado(CorreccionDictado correccion) {
  final pasos = <Paso>[];

  if (correccion.perfecto) {
    pasos.add(const Habla(Frases.todoBien));
  } else {
    final faltas = correccion.explicadas.take(maxFaltasARepasar).toList();
    pasos.add(Habla(Frases.resumenFallos(
      correccion.faltas,
      faltas.map((f) => f.esperado).toList(),
    )));
    for (final falta in faltas) {
      pasos.add(Habla(explicarPalabra(falta)));
    }
    if (correccion.faltas > maxFaltasARepasar) {
      pasos.add(const Habla(Frases.animo));
    }
  }

  if (!correccion.perfecto) pasos.add(const Habla(Frases.puedesRepetir));

  return Guion(
    asignatura: Asignatura.dictado,
    titulo: 'Repaso',
    pasos: pasos,
    comandosGlobales: const [Comando.repite, Comando.continua],
  );
}

// ------------------------------------------------------------ matemáticas ---

/// Una tanda de ejercicios: matemáticas o inglés.
///
/// Las dos funcionan igual —se plantea uno, el niño lo escribe y lo resuelve, y
/// se pasa al siguiente cuando él lo dice—, así que comparten guion. Lo único
/// que cambia es cómo se presenta y que un enunciado de inglés lleva dentro
/// trozos que hay que decir en inglés.
Guion guionTanda(Asignatura asignatura, List<Ejercicio> ejercicios) {
  final conProblemas = ejercicios.any((op) => op.tienePlanteamiento);
  final esIngles = asignatura == Asignatura.ingles;

  final pasos = <Paso>[
    Habla(esIngles
        ? Frases.inglesIntro(ejercicios.length)
        : Frases.matematicasIntro(ejercicios.length, conProblemas: conProblemas)),
    const Espera(Frases.prepararPapel, [Comando.listo]),
    const Habla(Frases.empezamosMates),
  ];

  for (final op in ejercicios) {
    pasos.add(Fragmento(
      indice: op.numero - 1,
      texto: Frases.ejercicioNumero(op.numero, op.dictado),
      // Un planteamiento con enunciado tampoco se retiene a la primera: se lee dos
      // veces, igual que una frase de dictado.
      veces: op.tienePlanteamiento ? 2 : 1,
      // Pero no palabra a palabra: esto no se copia, se entiende.
      palabraAPalabra: false,
      escrito: '${op.numero}.  ${op.planteamiento ?? op.enunciado}',
    ));
  }

  pasos.add(Revisar(esIngles
      ? Frases.inglesFin(ejercicios.length)
      : Frases.matematicasFin(ejercicios.length)));

  return Guion(
    asignatura: asignatura,
    titulo: switch (asignatura) {
      Asignatura.ingles => 'Inglés',
      _ => conProblemas ? 'Ejercicios' : 'Operaciones',
    },
    pasos: pasos,
    comandosGlobales: const [Comando.repite, Comando.masDespacio, Comando.continua],
  );
}

/// Repaso de una tanda ya corregida, sea de matemáticas o de inglés.
///
/// Con `modoPistas` la voz no da la solución de entrada: suelta una pista y
/// pregunta si ya lo ve. Solo si el niño pide una segunda pista y sigue sin
/// verlo se le da la respuesta. Es más lento y es el objetivo: quien corrige el
/// ejercicio tiene que ser el niño.
Guion guionRepasoTanda(
  Asignatura asignatura,
  List<ResultadoEjercicio> resultados,
  bool modoPistas,
) {
  final fallos = resultados.where((r) => !r.correcta).toList();
  final pasos = <Paso>[];

  if (fallos.isEmpty) {
    pasos.add(const Habla(Frases.todoBien));
  } else {
    pasos.add(Habla(fallos.length == 1
        ? Frases.casiTodoBien(1)
        : Frases.resumenFallos(
            fallos.length, fallos.map((f) => f.ejercicio.enunciado).toList())));

    for (final fallo in fallos.take(maxFaltasARepasar)) {
      final op = fallo.ejercicio;
      pasos.add(Habla(Frases.fallasteEn(op.numero)));

      if (!modoPistas) {
        pasos.add(Habla(
            '${Frases.solucion(op.respuestaEnVozAlta)} ${op.explicacion}'));
        continue;
      }

      pasos.add(Habla(op.pistas[0]));
      pasos.add(Pregunta(Frases.loTienes, [
        const RamaPregunta(
          Comando.loTengo,
          [Habla('Muy bien. Corrígela en el cuaderno.')],
        ),
        RamaPregunta(Comando.otraPista, [
          Habla(op.pistas[1]),
          Pregunta(Frases.loTienes, [
            const RamaPregunta(
              Comando.loTengo,
              [Habla('Eso es. Corrígela en el cuaderno.')],
            ),
            RamaPregunta(Comando.otraPista, [
              Habla('${Frases.solucion(op.respuestaEnVozAlta)} ${op.explicacion}'),
            ]),
          ]),
        ]),
      ]));
    }
    pasos.add(const Habla(Frases.puedesRepetir));
  }

  return Guion(
    asignatura: asignatura,
    titulo: 'Repaso',
    pasos: pasos,
    comandosGlobales: const [Comando.repite, Comando.continua],
  );
}

File: backend/src/domain/actividades.ts
import type { Correccion, ErrorDictado } from "../correccion/alinear.js";
import type { Dictado } from "../content/dictados.js";
import { pausaSegundos } from "../content/dictados.js";
import type { Operacion } from "../content/matematicas.js";
import { explicarFalta, frases } from "../tts/frases.js";
import { espera, fragmento, habla, pedirFoto, pregunta, type Guion, type Paso } from "./guion.js";

/** Cuántas faltas se repasan de viva voz antes de que el repaso canse más de lo que enseña. */
const MAX_FALTAS_A_REPASAR = 5;

// ---------------------------------------------------------------- dictado ---

export async function guionDictado(actividadId: string, dictado: Dictado): Promise<Guion> {
  const pasos: Paso[] = [
    await habla(frases.dictadoIntro(dictado.titulo)),
    await espera(frases.prepararPapel(), ["listo"], "Estoy listo"),
    await habla(frases.empezamos()),
  ];

  for (const [i, texto] of dictado.fragmentos.entries()) {
    pasos.push(await fragmento(i, texto, pausaSegundos(texto, dictado.nivel)));
  }

  pasos.push(await pedirFoto(frases.dictadoFin()));

  return {
    actividadId,
    asignatura: "dictado",
    titulo: dictado.titulo,
    // Durante un dictado el niño tiene que poder interrumpir sin tocar nada.
    comandosGlobales: ["repite", "mas_despacio", "mas_rapido", "continua"],
    pasos,
  };
}

export async function guionRepasoDictado(
  actividadId: string,
  correccion: Correccion,
): Promise<Guion> {
  const pasos: Paso[] = [];
  const faltas = correccion.errores.slice(0, MAX_FALTAS_A_REPASAR);

  if (correccion.errores.length === 0) {
    pasos.push(await habla(frases.todoBien()));
  } else {
    const palabras = faltas.map((e) => e.esperado).filter((p) => p !== "");
    pasos.push(await habla(frases.resumenFallos(correccion.errores.length, palabras)));
    for (const falta of faltas) {
      pasos.push(await habla(explicarFalta(falta)));
    }
    if (correccion.errores.length > MAX_FALTAS_A_REPASAR) {
      pasos.push(await habla(frases.animo()));
    }
  }

  return {
    actividadId,
    asignatura: "dictado",
    titulo: "Repaso",
    comandosGlobales: ["repite", "continua"],
    pasos,
  };
}

// ------------------------------------------------------------ matemáticas ---

/**
 * Cuánto callar tras dictar una operación. Es el tiempo de *copiarla*, no el de
 * resolverla: se dictan todas primero y el niño las resuelve después a su ritmo,
 * para que el móvil no esté hablándole mientras piensa.
 */
function pausaParaCopiar(op: Operacion): number {
  return Math.round(6 + op.enunciado.length * 0.6);
}

export async function guionMatematicas(
  actividadId: string,
  operaciones: readonly Operacion[],
): Promise<Guion> {
  const pasos: Paso[] = [
    await habla(frases.matematicasIntro(operaciones.length)),
    await espera(frases.prepararPapel(), ["listo"], "Estoy listo"),
    await habla(frases.empezamos()),
  ];

  for (const op of operaciones) {
    pasos.push(
      await fragmento(op.numero - 1, frases.operacion(op.numero, op.dictado), pausaParaCopiar(op)),
    );
  }

  pasos.push(await pedirFoto(frases.matematicasFin(operaciones.length)));

  return {
    actividadId,
    asignatura: "matematicas",
    titulo: "Operaciones",
    comandosGlobales: ["repite", "mas_despacio", "continua"],
    pasos,
  };
}

export interface OperacionCorregida {
  operacion: Operacion;
  escrito: string;
  correcta: boolean;
}

/**
 * Repaso de una tanda de operaciones.
 *
 * Con `modoPistas` la voz no da la solución de entrada: suelta una pista y
 * pregunta si ya lo ve. Solo si el niño pide una segunda pista y sigue sin
 * verlo, se le da la respuesta. Es más lento y es el objetivo: quien corrige el
 * ejercicio tiene que ser el niño.
 */
export async function guionRepasoMatematicas(
  actividadId: string,
  resultados: readonly OperacionCorregida[],
  modoPistas: boolean,
): Promise<Guion> {
  const fallos = resultados.filter((r) => !r.correcta);
  const pasos: Paso[] = [];

  if (fallos.length === 0) {
    pasos.push(await habla(frases.todoBien()));
  } else {
    pasos.push(
      await habla(
        fallos.length === 1
          ? frases.casiTodoBien(1)
          : frases.resumenFallos(fallos.length, fallos.map((f) => f.operacion.enunciado)),
      ),
    );

    for (const fallo of fallos.slice(0, MAX_FALTAS_A_REPASAR)) {
      const { operacion } = fallo;
      pasos.push(await habla(frases.fallasteEn(operacion.numero)));

      if (!modoPistas) {
        pasos.push(await habla(`${frases.solucion(operacion.respuesta)} ${operacion.explicacion}`));
        continue;
      }

      pasos.push(await habla(operacion.pistas[0]));
      pasos.push(
        await pregunta(frases.loTienes(), [
          {
            comando: "lo_tengo",
            etiqueta: "Ya lo veo",
            pasos: [await habla("Muy bien. Corrígela en el cuaderno.")],
          },
          {
            comando: "otra_pista",
            etiqueta: "Otra pista",
            pasos: [
              await habla(operacion.pistas[1]),
              await pregunta(frases.loTienes(), [
                {
                  comando: "lo_tengo",
                  etiqueta: "Ya lo veo",
                  pasos: [await habla("Eso es. Corrígela en el cuaderno.")],
                },
                {
                  comando: "otra_pista",
                  etiqueta: "Dímelo",
                  pasos: [
                    await habla(`${frases.solucion(operacion.respuesta)} ${operacion.explicacion}`),
                  ],
                },
              ]),
            ],
          },
        ]),
      );
    }
  }

  return {
    actividadId,
    asignatura: "matematicas",
    titulo: "Repaso",
    comandosGlobales: ["repite", "continua"],
    pasos,
  };
}

export type { ErrorDictado };
